### PR TITLE
History enhancements

### PR DESCRIPTION
**Pull request by diegogrz**
Logic changes to HistoryPropertyValue code to support.
#1013 Delete property event
#1035 Avoid modified events with no modifications (mostly InMemory implementation)

### CODE REVIEW

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java [hunk 1]
```diff
@@ -143,36 +160,52 @@ protected void deleteProperty(Property p) {
...
-                );
-                historicalPropertyValues.add(historicalPropertyValue);
+
+                builder.propertyVisibility(m.getVisibility())
+                        .timestamp(m.getTimestamp())
+                        .value(value)
+                        .metadata(apvm.getMetadata())
+                        .hiddenVisibilities(hiddenVisibilities)
+                        .isDeleted(false)
+                        .build();
```
Comment (mwizeman) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java]: It looks like this build is not needed.

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java [hunk 2]
```diff
@@ -143,36 +160,52 @@ protected void deleteProperty(Property p) {
...
+                    }
+                }
+
+                historicalPropertyValues.add(builder.build());
             } else {
                 throw new VertexiumException("Unhandled PropertyMutation: " + m.getClass().getName());
             }
         }
 
+
```
Comment (mwizeman) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java]: Please remove extra whitespace

File: core/src/main/java/org/vertexium/HistoricalPropertyValue.java [hunk 3]
```diff
@@ -3,32 +3,32 @@
...
-        this.propertyName = propertyName;
-        this.propertyVisibility = propertyVisibility;
-        this.timestamp = timestamp;
-        this.value = value;
-        this.metadata = metadata;
-        this.hiddenVisibilities = hiddenVisibilities;
+    private final String propertyKey;               // required
+    private final String propertyName;              // required
+    private final Visibility propertyVisibility;    // optional
+    private final long timestamp;                   // optional
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/HistoricalPropertyValue.java]: If this is optional should this be `Long` and set to `null`

File: core/src/main/java/org/vertexium/HistoricalPropertyValue.java [hunk 4]
```diff
@@ -3,32 +3,32 @@
...
+    private final Object value;                     // optional
+    private final Metadata metadata;                // optional
+    private final boolean isDeleted;                // optional
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/HistoricalPropertyValue.java]: Same here, should this be `Boolean` and set to `null`

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java [hunk 5]
```diff
@@ -143,36 +160,52 @@ protected void deleteProperty(Property p) {
...
             if (m instanceof SoftDeletePropertyMutation) {
-                continue;
+                builder.isDeleted(true);
+                historicalPropertyValues.add(builder.build());
             }
-            if (m instanceof AddPropertyMetadataMutation) {
-                continue;
+            else if (m instanceof AddPropertyMetadataMutation) {
+                builder.metadata(((AddPropertyMetadataMutation) m).getMetadata());
+                //historicalPropertyValues.add(builder.build());
```
Comment (joeferner) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java]: remove unused code

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java [hunk 6]
```diff
@@ -143,36 +160,52 @@ protected void deleteProperty(Property p) {
...
             }
+
             if (m instanceof SoftDeletePropertyMutation) {
-                continue;
+                builder.isDeleted(true);
+                historicalPropertyValues.add(builder.build());
             }
-            if (m instanceof AddPropertyMetadataMutation) {
-                continue;
+            else if (m instanceof AddPropertyMetadataMutation) {
```
Comment (joeferner) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java]: formatting. should be `} else if...` on one line

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java (hunk 2 above)
Comment (joeferner) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java]: remove extra whitespace

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java [hunk 7]
```diff
@@ -143,36 +160,52 @@ protected void deleteProperty(Property p) {
...
+
+                builder.propertyVisibility(m.getVisibility())
+                        .timestamp(m.getTimestamp())
+                        .value(value)
+                        .metadata(apvm.getMetadata())
+                        .hiddenVisibilities(hiddenVisibilities)
+                        .isDeleted(false)
+                        .build();
+
+                // Property modifications use a soft delete immediately followed by un add property mutation.
```
Comment (mwizeman) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryTableElement.java]: Comment typo: `un add property` should read `an add property`

**Comment by joeferner** [comment]
Looks good can you rebase and squash changes (`git rebase master -i`) then go ahead and merge.

**Pull request merged**